### ISSUE TITLE
Adding secret vars in config.json from exernal file

### ISSUE THREAD

**Issue by gdoumenc** [aws/chalice#1334]
I suggest to be able to add secret information via external file (as in terraform).

For example, something like that in ``config.json`` :

`"environment_variables": {"authorization_token": var.secret_token }`

And for the command line :

`$chalice deploy --var-file=secret.json`

**Comment by jamesls** [comment]
That would still mean that the env vars are sent to lambda unencrypted and anyone with read access can see your lambda vars.  Two ways that I've seen this done that I recommend:

* Use SSM parameter store, and retrieve those vars in your lambda function using the `ssm` client.
* Encrypt your env vars via `kms.encrypt` and then decrypt them in your lambda function at runtime.  In this case you can add them directly to your `.chalice/config.json` because they're encrypted.

For the second case, we could maybe add a command to automatically encrypt a value for you and add it to your env var config.  That way it's a pretty simple process to add secrets to your lambda function.  What do you think?

**Comment by danil-smirnov** [comment]
@jamesls I vote for this as well.
For me it's okay to have params visible in AWS console as access to Prod account should be strongly restricted anyway.
But I don't want to store secrets in repository so injecting them into config.json just before deployment would help a lot!

**Comment by gdoumenc** [comment]
I don't think we should encrypt the secret vars and adding them visible thru AWS access is not an issue as @danil-smirnov said. But simply having secret vars file in .gitignore will help a lot...

**Comment by jamesls** [comment]
If you don't care about the vars being in plain text, then the proposal in https://github.com/aws/chalice/issues/838 could work for you.  It would essentially let you override anything you want in the config file.  So for your example you'd have:

```
.chalice/config.json:

{ ... "environment_variables": {"auth_token": "PLACEHOLDER"}}

.chalice/config.local.json:

{... "environment_variables": {"auth_token": "actual-token"}}
```

and you'd have .chalice/config.local.json .gitignored.  The benefit of this approach is that it's more general that just env vars, you can override any part of the config you want locally without having to modify the actual `.chalice/config.json`.

As a future enhancement, we could also let you specify a local config if you didn't want the default value of `.chalice/config.local.json`, maybe `chalice deploy --config-override /path/to/local.json`.

What do you think?

**Comment by danil-smirnov** [comment]
@jamesls Thank you for the link, config.local.json idea seems superuseful to me and should be implemented with no doubt.
However, it's hard to use it as-is in my CI/CD (GitLab). GitLab CI supports key-value variables which are masked and could be injected securely into environment by deployment runner.
So if I can do `chalice deploy -e "my_var=${GITLAB_VAR}"` and have then my_var replaced in config.json by chalice it would be nice.
From the other side, I'm not sure if this should be chalice functionality or just external scripting using e.g. `envsubst`...

**Comment by gdoumenc** [comment]
@jamesis Thanks also for the link. I don't understand the need of a local configuration file as the `stage` option is dedicated to such purpose.
I agree with @danil-smirnov remark's : having dynamic environment variables is really usefull in CI/CD.

**Comment by jamesls** [comment]
@gdoumenc The need for a local configuration is so that you can override specific parts of your config that you *don't* want checked in to your repo.  In this specific example, it's secrets.

It's essentially what you're asking for, but in a more general way.  You want the ability to replace values in the `.chalice/config.json` with values from a separate file (`--var-file foo.json`) dynamically at runtime where `foo.json` isn't in your source repo.

What I'm proposing is that instead of having a custom syntax we'd have to parse out in every config value, e.g. `${var.myvalue}` and a new schema for this "var file", you instead provide a local chalice config file that we'd deep merge into the existing `.chalice/config.json`.  This is the "var file".

I'm not opposed to what's being proposed here.  There's a similar proposal from a while back that's more or less the same thing: https://github.com/aws/chalice/issues/608#issuecomment-345391098

**Comment by gdoumenc** [comment]
Right, providing a local config will give the expected solution. I agree. Thanks for yours answers.

**Comment by vfilimonov** [comment]
I would argue that storing credentials in config (even external, git-ignored) is not a good idea in general. Especially since you could use AWS ecosystem for that.

If not using AWS Secrets Manager, that could be expensive if you have a lot of secrets, you could easily place them in AWS SSM Parameter Store (and encrypt with default or dedicated key). And in the python code retrieving them is no more complicated than doing
`os.environ['my_secret']`. Namely:
```python
boto3.client('ssm').get_parameter(Name=`my_secret`, WithDecryption=True)['Parameter']
```
And you need to specify permissions in `config-prod.json`:
```json
        {
            "Effect": "Allow",
            "Action": ["ssm:GetParameter"],
            "Resource": [
              "arn:aws:ssm:<region>:<account>:parameter/my_secret"
            ]
        }
```

Other benefits of this approach are:
- easy rotation of secrets
- granular permissions to secrets/group of secrets

**Comment by jamesls** [comment]
@vfilimonov yes, that's usually what we do for our chalice apps.  You might run into rate limit issues depending on heavily your lambda function is invoked but otherwise has worked well for us.

The only thing I wish were a little easier is the management of the actual SSM params.  Is there a tool or anything else you use to manage the deployment of these SSM params?  Perhaps chalice could make this more streamlined for people that want to go the SSM approach.

**Comment by vfilimonov** [comment]
@jamesls unfortunately I don't know of such tools. I am using a script that takes YAML and uploads it to SSM ParamStore in order to automate the process. 

It might be useful for chalice to have some basic support for parameters, but I'm afraid that it might become a headache at some point, trying to wrap around of the corner cases (in which format to store parameters / types / key ids, should it be deleted on `chalice delete`, how to rotate them without redeploy, what if parameter has to be shared between chalice projects, etc.). After all chalice was not meant to be a swiss-knife.

Maybe easier to implement would be the functionality to execute user-scripts before/after deploy/delete - then environment could be easily set by external scripts.

**Comment by devangmehta123** [comment]
@jamesls , the last comment on this is from Mar 25. I generally agree with @vfilimonov 's comments on this, that using AWS tools is a better way to do this.

But if someone wants to keep their Lambda setup simple, like I do, then there is still value in allowing config.json to be able to dynamically set up environment variables to configure inside the Lambda at deploy time.

That would let some configurations flow from GitHub actions or GitLab instead of relying on an extra layer of configuration inside SSM.

It is helpful to those who run the entire test setup on Docker containers inside the CI/CD sever, outside of the AWS ecosystem, for example.

For example, all I would like here for 'chalice deploy' to copy some behaviour of docker-compose. In docker-compose, if you leave the value of an environment variable empty, it will try to get the value from the shell. And so config.json could just look like this.

Note that the injected key is just a list and you could default the value to empty if nothing is obtained from the shell.

```
"environment_variables": {
    "USED_IN_ALL_ENVS": "this_value_needed_everywhere"
}
"environment_variables_injected": [
    "INJECTED_VARIABLE_FROM_SHELL_INVOKING_CHALICE_DEPLOY"
]
```

I think that even a simple enhancement like that would certain workflows which want to keep it simple.

**Comment by theist** [comment]
What I'm doing at the moment is to have an ignored  `.env` file in `./chalicelib/files/` and load it using `dotenv` package.

But still is possible to deploy the lambda without this file if it is shared by a team. I have It checked by a small deploy script that looks for everything in place before calling `chalice deploy` and this gave me this idea:

What about allow chalice user to prehook into deploy using `app.py` special code? this way someone could use whatever it wants to sync secrets. from just pulling it from OS environment and build config.json using a template to get special files from a Vault server or whatever.
